src/modules-load/modules-load.c

```diff
@@ -39,45 +81,251 @@ static int parse_proc_cmdline_item(const char *key, const char *value, void *dat
         return 0;
 }
 
-static int apply_file(struct kmod_ctx *ctx, const char *path, bool ignore_enoent) {
+static int apply_file(FILE *f, const char *filename, OrderedSet *module_set) {
+        int ret = 0, r;
+
+        assert(f);
+
+        log_debug("apply: %s", filename);
+        for (;;) {
+                _cleanup_free_ char *line = NULL;
+
+                r = read_stripped_line(f, LONG_LINE_MAX, &line);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to read file '%s': %m", filename);
+                if (r == 0)
+                        break;
+
+                if (isempty(line))
+                        continue;
+                if (strchr(COMMENTS, *line))
+                        continue;
+
+                r = modules_list_append_take(module_set, &line);
+                if (r < 0)
+                        RET_GATHER(ret, log_error_errno(r, "Failed to add module '%s': %m", line));
+        }
+
+        return ret;
+}
+
+static int apply_file_from_path(const char *path, OrderedSet *module_set) {
         _cleanup_fclose_ FILE *f = NULL;
         _cleanup_free_ char *pp = NULL;
         int r;
 
-        assert(ctx);
         assert(path);
 
         r = search_and_fopen_nulstr(path, "re", NULL, conf_file_dirs, &f, &pp);
-        if (r < 0) {
-                if (ignore_enoent && r == -ENOENT)
+        if (r < 0)
+                return log_error_errno(r, "Failed to open %s: %m", path);
+
+        return apply_file(f, pp, module_set);
+}
+
+static int apply_conf_file(ConfFile *c, OrderedSet *module_set) {
+        _cleanup_fclose_ FILE *f = NULL;
+
+        f = fopen(FORMAT_PROC_FD_PATH(c->fd), "re");
+        if (!f) {
+                if (errno == ENOENT)
                         return 0;
 
-                return log_error_errno(r, "Failed to open %s: %m", path);
+                return log_error_errno(errno, "Failed to open %s: %m", c->original_path);
+        }
+
+        return apply_file(f, c->original_path, module_set);
+}
+
+static int do_direct_probe(OrderedSet *module_set) {
+        _cleanup_(sym_kmod_unrefp) struct kmod_ctx *ctx = NULL;
+        char *module;
+        int ret = 0, r;
+
+        r = module_setup_context(&ctx);
+        if (r < 0)
+                return log_error_errno(r, "Failed to initialize libkmod context: %m");
+
+        ORDERED_SET_FOREACH(module, module_set) {
+                r = module_load_and_warn(ctx, module, true);
+                if (r != -ENOENT)
+                        RET_GATHER(ret, r);
         }
 
-        log_debug("apply: %s", pp);
+        return ret;
+}
+
+static int enqueue_module_to_load(int sock, const char *module) {
+        ssize_t bytes;
+
+        assert(sock >= 0);
+        assert(module);
+
+        if (strlen(module) > MODULE_NAME_MAX_LEN) {
+                log_error("Module name max length exceeded (%lu): %s", MODULE_NAME_MAX_LEN, module);
+                return -1;
```

Comment:
please do not make up error codes. just return some sensible error here. i.e. EINVAL or so.

also use `return log_error_errno(SYNTHETIC_ERRNO(…)), …)`